Add chapter about CPU features dispatching into docs (#2945)
oneapi-src/oneDAL: docs/source/contribution/cpu_features.rst

.. ******************************************************************************

> .. * Copyright contributors to the oneDAL project

Reviewer: Question: I see other files always mention the year in the copyright notice - for example:
https://oneapi-src.github.io/oneDAL/examples/cpp/basic_statistics_dense_batch.html

Isn't it required to have it?

set architectures.

Following architectures are currently supported:

- Intel |reg| Streaming SIMD Extensions 2 (Intel |reg| SSE2)
- Intel |reg| Streaming SIMD Extensions 4.2 (Intel |reg| SSE4.2)
- Intel |reg| Advanced Vector Extensions 2 (Intel |reg| AVX2)
- Intel |reg| Advanced Vector Extensions 512 (Intel |reg| AVX-512)

The particular code path is chosen at runtime based on the underlying hardware characteristics.

This chapter describes how the code is organized to support this variety of instruction sets.

Algorithm Implementation Options
********************************

In addition to the instruction set architectures, an algorithm in oneDAL may have various
implementation options. Below is a description of these options to help you better understand
the oneDAL code structure and conventions.

Computational Tasks
-------------------

An algorithm might have various tasks to compute. The most common options are:

- `Classification <https://oneapi-src.github.io/oneDAL/onedal/glossary.html#term-Classification>`_,
- `Regression <https://oneapi-src.github.io/oneDAL/onedal/glossary.html#term-Regression>`_.

Computational Stages
--------------------

> An algorithm might have ``training`` and ``inference`` computaion stages aimed

Reviewer: ```suggestion
An algorithm might have ``training`` and ``inference`` computation stages aimed
```

> to train a model on the input dataset and compute the inference results respectively.

Reviewer: ```suggestion
at training a model on the input dataset and computing the inference results, respectively.
```

Computational Methods
---------------------

An algorithm can support several methods for the same type of computations.
For example, kNN algorithm supports
`brute_force <https://oneapi-src.github.io/oneDAL/onedal/algorithms/nearest-neighbors/knn.html#knn-t-math-brute-force>`_
and `kd_tree <https://oneapi-src.github.io/oneDAL/onedal/algorithms/nearest-neighbors/knn.html#knn-t-math-kd-tree>`_
methods for algorithm training and inference.

Computational Modes
-------------------

oneDAL can provide several computaional modes for an algorithm.
See `Computaional Modes <https://oneapi-src.github.io/oneDAL/onedal/programming-model/computational-modes.html>`_
chapter for details.

Folders and Files
*****************

> Consider you are working on some algorithm ``Abc`` in oneDAL.

Reviewer: ```suggestion
Suppose that you are working on some algorithm ``Abc`` in oneDAL.
```

The part of the implementation of this algorithms that is running on CPU should be located in
`cpp/daal/src/algorithms/abc` folder.

> Consider it provides:

Reviewer: ```suggestion
Suppose that it provides:
```

- ``classification`` and ``regression`` learning tasks;
- ``training`` and ``inference`` stages;
- ``method1`` and ``method2`` for the ``training`` stage and only ``method1`` for ``inference`` stage;
- only batch computational mode.

Then the `cpp/daal/src/algorithms/abc` folder should contain at least the following files:

::

  cpp/daal/src/algorithms/abc/

>     |-- abc_classification_predict_method1_batch_fpt_cpu.cpp

Reviewer: What does "fpt" stand for BTW?

Author: "floating point type" I guess.

Reviewer: Thanks. Perhaps it could be added as a small comment.

Author: The description was added in '*_fpt_cpu.cpp' subsection.

    |-- abc_classification_predict_method1_impl.i
    |-- abc_classification_predict_kernel.h
    |-- abc_classification_train_method1_batch_fpt_cpu.cpp
    |-- abc_classification_train_method2_batch_fpt_cpu.cpp
    |-- abc_classification_train_method1_impl.i
    |-- abc_classification_train_method2_impl.i
    |-- abc_classification_train_kernel.h
    |-- abc_regression_predict_method1_batch_fpt_cpu.cpp
    |-- abc_regression_predict_method1_batch_fpt_cpu.cpp
    |-- abc_regression_predict_method1_impl.i
    |-- abc_regression_predict_kernel.h

>     |-- abc_regression_train_method1_batch_fpt_cpu.cpp

Author: Those `_dispatcher.cpp` files and also `_container.h` and some other files are part of legacy DAAL API.
They are not related to CPU features dispatching in oneDAL, so I am not touching those in the scope of this PR.

>     |-- abc_regression_train_method1_batch_fpt_cpu.cpp

Author: They are not yet deprecated. But we do not add new DAAL API, we add oneDAL API instead.

Yes, I can add the note that other algorithms folders could contain additional files, not described in this chapter, most of them are related to DAAL interface (which we are not going to extend) and are not related to oneDAL or CPU features support.

    |-- abc_regression_train_method2_batch_fpt_cpu.cpp
    |-- abc_regression_train_method1_impl.i
    |-- abc_regression_train_method2_impl.i
    |-- abc_regression_train_kernel.h

> Alternative variant of the folder structure to avoid storing too much files within a single folder

Reviewer: ```suggestion
Alternative variant of the folder structure to avoid storing too many files within a single folder
```

> can be:

Reviewer: ```suggestion
could be:
```

          |-- abc_regression_train_method2_batch_fpt_cpu.cpp
          |-- abc_regression_train_method1_impl.i
          |-- abc_regression_train_method2_impl.i
          |-- abc_regression_train_kernel.h

The names of the files stay the same in this case, just the folder layout differs.

Further the purpose and contents of each file are to be described on the example of classification
training task. For other types of the tasks the structure of the code is similar.

\*_kernel.h
-----------

Those files contain the definitions of one or several template classes that define member functions that
do the actual computations. Here is a variant of the ``Abc`` training algorithm kernel definition in the file
`abc_classification_train_kernel.h`:

.. include:: ../includes/cpu_features/abc-classification-train-kernel.rst

Typical template parameters are:

- ``algorithmFPType``  Data type to use in intermediate computations for the algorithm,
                       ``float`` or ``double``.
- ``method`` Computational methods of the algorithm. ``method1`` or ``method2`` in the case of ``Abc``.
- ``cpu`` Version of the cpu-specific implementation of the algorithm, ``daal::CpuType``.

> Implementations for different methods are usually defined usind partial class templates specialization.

Author: Fixed.

This will guide the compiler to generate more efficient code for the target architecture.

Consider that the implementation of the ``method2`` for the same algorithm will be different and will contain
AVX-512-specific code located in ``cpuSpecificCode`` function.
Then the implementation of the ``method2`` in the file `abc_classification_train_method2_impl.i` will look like:

.. include:: ../includes/cpu_features/abc-classification-train-method2-impl.rst

CPU-specific code needs to be placed under compiler-specific and CPU-specific defines because it usually
contains intrinsics that cannot be compiled on other architectures.

\*_fpt_cpu.cpp
--------------

Those files contain the instantiations of the template classes defined in `*_kernel.h` files.
The instatiation of the ``Abc`` training algorithm kernel for ``method1`` is located in the file
`abc_classification_train_method1_batch_fpt_cpu.cpp`:

.. include:: ../includes/cpu_features/abc-classification-train-method1-fpt-cpu.rst

`_fpt_cpu.cpp` files are not compiled directly into object files. First, multiple copies of those files

> are made raplacing the ``fpt`` and ``cpu`` parts of the file name as well as the corresponding ``DAAL_FPTYPE`` and

Reviewer: ```suggestion
are made replacing the ``fpt`` and ``cpu`` parts of the file name as well as the corresponding ``DAAL_FPTYPE`` and
```

> ``DAAL_CPU`` macros with the actual data type and CPU type values. Then the resulting files are compiled

Reviewer: Perhaps it could mention here that SIMD code also needs to be wrapped into an `#ifdef` section with conditions like `DAAL_INTEL_CPP_COMPILER && (__CPUID__(DAAL_CPU) == __avx512__)`, in addition to having a templated version.

with appropriate CPU-specific optimization compiler options.

The values for ``fpt`` file name part replacement are:
- ``flt`` for ``float`` data type, and
- ``dbl`` for ``double`` data type.

The values for ``DAAL_FPTYPE`` macro replacement are ``float`` and ``double`` respectively.

The values for ``cpu`` file name part replacement are:

> - ``nrh`` for Intel |reg| SSE2 architecture, which stands for Northwood,

Reviewer: What's the difference between these and `DAAL_CPU`? Is there some case in which they wouldn't have a 1-to-1 match? When should these be preferred over `DAAL_CPU`?

Author: There is really no big difference. They are basically describe the same parameter of the resulting product.
`cpu` is a part of the file names - matters to build system, and `DAAL_CPU` is a macro which is a part of the code.
Of course, it is possible to have those two (`cpu` and `DAAL_CPU`) in alignment. But this chapter describes how it is done now.

I do not know exactly, but it might be that those two values were made different intentionally.
I can imagine two different CPUs that share the same instruction set, but have different througputs/latencies of the instructions; so the compiler might provide different optimization option for those CPUs.
In this case it would be beneficial to differentiate between `cpu` and `DAAL_CPU` variables.

.. *
.. * Licensed under the Apache License, Version 2.0 (the "License");
.. * you may not use this file except in compliance with the License.
.. * You may obtain a copy of the License at
.. *
.. *     http://www.apache.org/licenses/LICENSE-2.0
.. *
.. * Unless required by applicable law or agreed to in writing, software
.. * distributed under the License is distributed on an "AS IS" BASIS,
.. * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
.. * See the License for the specific language governing permissions and
.. * limitations under the License.
.. *******************************************************************************/

.. highlight:: cpp

CPU Features Dispatching
^^^^^^^^^^^^^^^^^^^^^^^^

For each algorithm |short_name| provides several code paths for x86-64-compatibe instruction
set architectures.

Following architectures are currently supported:

- Intel\ |reg|\  Streaming SIMD Extensions 2 (Intel\ |reg|\  SSE2)
- Intel\ |reg|\  Streaming SIMD Extensions 4.2 (Intel\ |reg|\  SSE4.2)
- Intel\ |reg|\  Advanced Vector Extensions 2 (Intel\ |reg|\  AVX2)
- Intel\ |reg|\  Advanced Vector Extensions 512 (Intel\ |reg|\  AVX-512)

> The particular code path is chosen at runtime based on the underlying hardware characteristics.

Reviewer 2: ```suggestion
The particular code path is chosen at runtime based on underlying hardware properties.
```

    |     |-- abc_classification_predict_method1_batch_fpt_cpu.cpp
    |     |-- abc_classification_predict_method1_impl.i
    |     |-- abc_classification_predict_kernel.h
    |     |-- abc_classification_train_method1_batch_fpt_cpu.cpp
    |     |-- abc_classification_train_method2_batch_fpt_cpu.cpp
    |     |-- abc_classification_train_method1_impl.i
    |     |-- abc_classification_train_method2_impl.i
    |     |-- abc_classification_train_kernel.h
    |-- regression/
          |-- abc_regression_predict_method1_batch_fpt_cpu.cpp
          |-- abc_regression_predict_method1_impl.i
          |-- abc_regression_predict_kernel.h
          |-- abc_regression_train_method1_batch_fpt_cpu.cpp
          |-- abc_regression_train_method2_batch_fpt_cpu.cpp
          |-- abc_regression_train_method1_impl.i
          |-- abc_regression_train_method2_impl.i
          |-- abc_regression_train_kernel.h

The names of the files stay the same in this case, just the folder layout differs.

Further the purpose and contents of each file are to be described on the example of classification
training task. For other types of the tasks the structure of the code is similar.

\*_kernel.h
-----------

> Those files contain the definitions of one or several template classes that define member functions that

Reviewer 2: nit: Don't start the section with a pronoun. Put a full description of what you are describing. Maybe:

```
In the directory structure introduced in the last section, there are files with a `_kernel.h` suffix. These contain the definitions of ...
```

do the actual computations. Here is a variant of the ``Abc`` training algorithm kernel definition in the file
`abc_classification_train_kernel.h`:

.. include:: ../includes/cpu_features/abc-classification-train-kernel.rst

Typical template parameters are:

> - ``algorithmFPType``  Data type to use in intermediate computations for the algorithm,
>                        ``float`` or ``double``.

Reviewer 2: ```suggestion
- ``algorithmFPType``  Data type to use in intermediate computations for the algorithm.
                       Must be one of ``float`` or ``double``.
```

- ``method`` Computational methods of the algorithm. ``method1`` or ``method2`` in the case of ``Abc``.
- ``cpu`` Version of the cpu-specific implementation of the algorithm, ``daal::CpuType``.

Implementations for different methods are usually defined using partial class templates specialization.

\*_impl.i
---------

> Those files contain the implementations of the computational functions defined in `*_kernel.h` files.

Reviewer 2: nit: Don't start the section with a pronoun. See the similar comment at the start of the `\*_kernel.h` section

Here is a variant of ``method1`` imlementation for ``Abc`` training algorithm that does not contain any
instruction set specific code. The implementation is located in the file `abc_classification_train_method1_impl.i`:

.. include:: ../includes/cpu_features/abc-classification-train-method1-impl.rst

Although the implementation of the ``method1`` does not contain any instruction set specific code, it is
expected that the developers leverage SIMD related macros available in |short_name|.

> For example, ``PRAGMA_IVDEP``, ``PRAGMA_VECTOR_ALWAYS``, ``PRAGMA_VECTOR_ALIGNED`` and others pragmas defined in

Reviewer 2: ```suggestion
For example, ``PRAGMA_IVDEP``, ``PRAGMA_VECTOR_ALWAYS``, ``PRAGMA_VECTOR_ALIGNED`` and other pragmas defined in
```

`service_defines.h <https://github.com/oneapi-src/oneDAL/blob/main/cpp/daal/src/services/service_defines.h>`_.
This will guide the compiler to generate more efficient code for the target architecture.

Consider that the implementation of the ``method2`` for the same algorithm will be different and will contain
AVX-512-specific code located in ``cpuSpecificCode`` function. Note that all the compiler-specific code should

> be placed under compiler-specific defines. For example, the Intel\ |reg|\  oneAPI DPC++/C++ Compiler specific code
> should be placed under ``DAAL_INTEL_CPP_COMPILER`` define. All the CPU-specific code should be placed under
> CPU-specific defines. For example, the AVX-512 specific code should be placed under
> ``__CPUID__(DAAL_CPU) == __avx512__``.

Reviewer 2: ```suggestion
be gated by values of compiler-specific defines. For example, the Intel\ |reg|\  oneAPI DPC++/C++ Compiler specific code
should be gated by the existence of the ``DAAL_INTEL_CPP_COMPILER`` define. All the CPU-specific code should be gated on the value of
CPU-specific defines. For example, the AVX-512 specific code should be gated on the value
``__CPUID__(DAAL_CPU) == __avx512__``.
```

Then the implementation of the ``method2`` in the file `abc_classification_train_method2_impl.i` will look like:

.. include:: ../includes/cpu_features/abc-classification-train-method2-impl.rst

\*_fpt_cpu.cpp
--------------

> Those files contain the instantiations of the template classes defined in `*_kernel.h` files.

Reviewer 2: nit: Don't start a section with a pronoun. See the similar comment at the start of the `\*_kernel.h` section

The instatiation of the ``Abc`` training algorithm kernel for ``method1`` is located in the file
`abc_classification_train_method1_batch_fpt_cpu.cpp`:

.. include:: ../includes/cpu_features/abc-classification-train-method1-fpt-cpu.rst

`_fpt_cpu.cpp` files are not compiled directly into object files. First, multiple copies of those files
are made replacing the ``fpt``, which stands for 'floating point type', and ``cpu`` parts of the file name
as well as the corresponding ``DAAL_FPTYPE`` and ``DAAL_CPU`` macros with the actual data type and CPU type values.

> Then the resulting files are compiled with appropriate CPU-specific optimization compiler options.

Reviewer 2: ```suggestion
Then the resulting files are compiled with appropriate CPU-specific compiler optimization options.
```

The values for ``DAAL_FPTYPE`` macro replacement are ``float`` and ``double``, respectively.

The values for ``cpu`` file name part replacement are:

- ``nrh`` for Intel\ |reg|\  SSE2 architecture, which stands for Northwood,
- ``neh`` for Intel\ |reg|\  SSE4.2 architecture, which stands for Nehalem,
- ``hsw`` for Intel\ |reg|\  AVX2 architecture, which stands for Haswell,
- ``skx`` for Intel\ |reg|\  AVX-512 architecture, which stands for Skylake-X.

The values for ``DAAL_CPU`` macro replacement are:

- ``__sse2__`` for Intel\ |reg|\  SSE2 architecture,
- ``__sse42__`` for Intel\ |reg|\  SSE4.2 architecture,
- ``__avx2__`` for Intel\ |reg|\  AVX2 architecture,
- ``__avx512__`` for Intel\ |reg|\  AVX-512 architecture.

Build System Configuration
**************************

This chapter describes which parts of the build system need to be modified to add new architectural
extensions to the build system or to remove an outdated one.

Makefile
--------

The most important definitions and functions for CPU features dispatching are located in the files
|32e_make|_ for x86-64 architecture, |riscv_make|_ for RISC-V 64-bit architecture, and |arm_make|_
for ARM architecture.

> Those files are included into operating system related files.

Reviewer: ```suggestion
Those files are included into operating system related makefiles.
```

For example, the |32e_make| file is included into ``lnx32e.mk`` file:

::

  include dev/make/function_definitions/32e.mk

And ``lnx32e.mk`` and similar files are included into the main Makefile:

::

  include dev/make/function_definitions/$(PLAT).mk

Where ``$(PLAT)`` is the platform name, for example, ``lnx32e``, ``win32e``, ``lnxriscv64``, etc.

To add a new architectural extension into |32e_make| file, ``CPUs`` and ``CPUs.files`` lists need to be updated.
The functions like ``set_uarch_options_for_compiler`` and others should also be updated accordingly.

> The compiler options for the new architectural extension should be added to the respective file in

Reviewer: ```suggestion
The compiler options for the new architectural extension should be added to the respective file in the
```
